README.md
Create image with multiple sizes based on [intervention](http://image.intervention.io/) easily.

*for Example:*

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::make($request->image)
  ->setExclusiveDirectory('post')
  ->save();
```

It will save your image in three sizes, which were defined in config file, in path:

- ```public/``` 
  - ```images/post/2021/12/2/1638611107/```
    - ```1638611107_960_large.png```
    - ```1638611107_960_medium.png```
    - ```1638611107_960_small.png```



## Prerequisites

- Laravel ```^8.0|^9.0```
- PHP ```^7.4|^8.0```

## Installation

```bash
composer require amir-hossein5/laravel-image
```

and for publishing configuration file: 

```bash
php artisan vendor:publish --tag image
```

And finally you may specify your sizes in configuration:

```
'use_size' => 'imageSizes',

 'imageSizes' => [
    'large' => [
        'width' => '800',
        'height' => '600'
    ],
    'medium' => [
        'width' => '400',
        'height' => '300'
    ],
    'small' => [
        'width' => '80',
        'height' => '60'
    ]
],
``` 


 
## "make" method

In ```make``` method defaults for directories(archive path(2021/12/2)) and sizes will be set. Beacause of this, when you use:

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::make($request->image) // or pass **Intervention object**
  ->setExclusiveDirectory('post')
  ->save();
```

saves in:

- ```public/``` 
  - ```images/post/2021/12/2/1638611107/```
    - ```1638611107_960_large.png```
    - ```1638611107_960_medium.png```
    - ```1638611107_960_small.png```


It will create your image in archive path(2021/12/2), and with those sizes, that you defined in config (in ```use_size``` part). 
But how to customize directories, and sizes.

For size customazations see [Size customazations](#size-customazations).

### Directory customazations

In parantheses written name of each directory:


- ```images(rootDirectory)```
- ```/post(exclusiveDirectory)```
- ```/2021/12/2/(archiveDirectories)```
- ```/1638611107(sizesDirectory)``` -> **if there be more than one size**

Image path setters:

| setter                          | default                                              |
|---------------------------------|------------------------------------------------------|
| setImageName( string )          | time()_rand(100, 999)_sizeName(if there be any size) |
| setImageFormat( string )        | uploaded image format.  e.g, ```->setImageFormat('png')``` |
| be( string )                    | sets both image name and format. e.g, ```->be('name.png')```| 
| setSizesDirectory( string )     | time()                                               |

Just available for ```make```:

| setter                          | default                                              |
|---------------------------------|------------------------------------------------------|
| setRootDirectory( string )      | images (written in config file)                      |
| setExclusiveDirectory( string ) |                                                      |
| setArchiveDirectories( string ) | year/month/day                                       |


> Notice: root directory is also changeable in config file.

Example: 

```php 
use AmirHossein5\LaravelImage\Facades\Image;

Image::make($image)
  ->setExclusiveDirectory('book')
  ->setRootDirectory('image')
  .
  .
  .
  ->save()
```

For size customazations see [Size customazations](#size-customazations).


## "raw" method

When you are using "raw" method,

nothing will be automatically set(archive path(2021/12/2), and sizes). For setting directory of image:

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::raw($image) // or pass **Intervention object**
  ->in('book')
  ->save()
  
// without resizing
// will be save in public/book/ 

Image::raw($image)->save()
  
// without resizing
// will be save in public/
```

For add size, and size customazations see [Size customazations](#size-customazations).

## Save in storage

For saving into storage folder you can use disks, which defined in config:

```php
'disks' => [
    'public' => public_path(),
    'storage' => storage_path('app')
]
```

By default it's ```public```.

```php
use AmirHossein5\LaravelImage\Facades\Image;

$images = Image::make($image)
  ->setExclusiveDirectory('post')
  ->disk('storage')
  ->save();
```

Will be:

- ```storage/app/``` 
  - ```images/post/2021/12/2/1638611107/```
    - ```1638611107_960_large.png```
    - ```1638611107_960_medium.png```
    - ```1638611107_960_small.png```

You may add more disks and use from that.



## Size customazations

Size setters:

| setter                                    | description                                                                         |
|-------------------------------------------|-------------------------------------------------------------------------------------|
| autoResize()                              | removes previous defined sizes                                                      |
| resize( $width, $height, $as = null )     | removes previous defined sizes, and adds a size                                     |
| alsoResize( $width, $height, $as = null ) | adds a size                                                                         |
| resizeBy( array )                         | resize by intended array(the structure shuold be like 'imageSizes' in configuration)|

> For automatically resizing use ```make``` method and define sizes in config file (in ```use_size``` part).


### Default size

If you want to use default_size functionality, you may define it in config file, or after image was created:

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::setDefaultSizeFor($post->image, 'small');
```

Will return previous array but default_size has changed or added.

<!-- 
If you created your result array **manually** pass the key of array, which there is image path(s):

```php
use AmirHossein5\LaravelImage\Facades\Image;

$image = Image::make($this->image)
  ->setExclusiveDirectory('post')
  ->save(false, function ($image) {
    return [
      'paths' => $image->imagePath,
    ];
  });
```

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::setDefaultSizeFor($post->image, 'small', 'paths');
```
 -->



## Result array

After creating image if operation was successful, it returns array, which

```index``` key is array,or string(contains one, or more) image paths, which depends on number of sizes.

For example:

```php
[
  "index" => [
      "images/post/2021/12/08/1638966454/1638966454_491_large.png",
      "images/post/2021/12/08/1638966454/1638966454_491_meduim.png",
      "images/post/2021/12/08/1638966454/1638966454_491_small.png",
   ]
   // or 
   "index" => "image path"
   
   "imageDirectory" =>  "images/post/2021/12/08/1638966454"
   "default_size" => 'medium' 
   "disk" => 'public'
]
```
> ```default_size``` key when you are using "default_size", and there are **more than one size**, will be.



### Getting result array manually

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::make($this->image)
  ->setExclusiveDirectory('post')
  ->save(false, function ($image) {
    return [
      'index' => $image->imagePath,
      'imageDirectory' => $image->imageDirectory,
      'disk' => $image->disk,
    ];
  });
```

output:

``` 
[
  "index" => [
      "images/post/2021/12/08/1638966454/1638966454_491_large.png",
      "images/post/2021/12/08/1638966454/1638966454_491_meduim.png",
      "images/post/2021/12/08/1638966454/1638966454_491_small.png",
   ]
   // or 
   "index" => "images/post/2021/12/08/1638966454/"
   
   "imageDirectory" => "images/post/2021/12/08/1638966454"
   "disk" => 'public'
]
```
  


Properties:

| Property | Description |
|----------------------------|-----------------------------------------------------------|
| $image->image              |    Uploaded image object.                                 |
| $image->sizes              | All used sizes.                                           |
| $image->defaultSize        | Default size.                                             |
| $image->imagePath          | Full path of stored image(s).                             |
| $image->imageDirectory     | Image's directory.                                        |
| $image->imageName          | Image name.                                               |
| $image->imageFormat        | Image format.                                             |
| $image->disk               | [Disk](#save-in-storage)                                  |
| $image->quality               | [Image Quality](#image-quality)                                  |
| $image->rootDirectory      | [see Directory customazations](#directory-customazations) |
| $image->exclusiveDirectory | [see Directory customazations](#directory-customazations) |
| $image->archiveDirectories | [see Directory customazations](#directory-customazations) |
| $image->sizesDirectory     | [see Directory customazations](#directory-customazations) |


## Upsize or not

If you want to use upsize of intervention you should:

```php
  ->save(true)
```

## Removeing image(s)

Pass created image:

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::rm($post->image);
```
which returns ```true``` or ```false```.

</br>

If you created your result array **manually** pass the key of array, which there is image path(s):

```php
use AmirHossein5\LaravelImage\Facades\Image;

$image = Image::make($this->image)
  ->setExclusiveDirectory('post')
  ->save(false, function ($image) {
    return ['paths' => $image->imagePath];
  });
  
Image::rm($image, 'paths');

```

Or if it's one string path just pass it:

```php
use AmirHossein5\LaravelImage\Facades\Image;

$image = Image::raw($this->image)
  ->in('post/test')
  ->save(false, function ($image) {
    return $image->imagePath;
  });
  
Image::rm($image);
```

to check that is removed, or not:

```php
use AmirHossein5\LaravelImage\Facades\Image;

if (! Image::rm($image)) {
  // ...
}
```

```php
use AmirHossein5\LaravelImage\Facades\Image;

if (! Image::wasRecentlyRemoved()) {
  // ...
}
```

#### Remove when used disk

If you created image with some disks do:

```php
use AmirHossein5\LaravelImage\Facades\Image;

....
  ->disk('storage')
....
  
Image::disk('storage')->rm($image);
```



## Replacing image(s)

```replace``` method works same as ```save``` method, but if there be image(s) with same name as this image, this will be replace.

```php
use AmirHossein5\LaravelImage\Facades\Image;

$image = Image::raw($this->image)
    ->be('logo.png')
    ->replace(); 
```

It works for multipe images too.


## Image Quality

It is normalized for all file types to a range from 0 (poor quality, small file) to 100 (best quality, big file). Quality is only applied if you're encoding JPG format since PNG compression is lossless and does not affect image quality. The default value is 90.

```php
  ->quality(90)
```

## Transactions

If an exception is thrown within the transaction closure, the transaction will automatically be rolled back and the exception is re-thrown. If the closure executes successfully, the transaction will automatically be committed.

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::transaction(function () {
  Image::raw($image)->save();
});
```

### Max Attempts
The ```transaction``` method accepts an optional second argument which defines the number of times a transaction should be retried when a deadlock occurs.

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::transaction(function () {
  Image::raw($image)->save();
}, maxAttempts: 4);
```

### Manually Using Transactions

For begin transaction manually:
```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::beginTransaction();
```
You can rollback the transaction via the ```rollBack``` method:
```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::rollBack();
```
Lastly, you can commit a transaction via the ```commit``` method:
```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::commit();
```


## Testing

If you don't want to image be ceate use fake before your code:

```php
use AmirHossein5\LaravelImage\Facades\Image;

Image::fake();

Image:: ...
```


## Examples

```php
use AmirHossein5\LaravelImage\Facades\Image;

$request['image'] = Image::make($request['image'])
  ->setExclusiveDirectory('post')
  ->save();

if (!$request['image']) {
  return back()
    ->withInput()
    ->withErrors(['image' => __('validation.uploaded')]);
}

Post::create($request);
```

```php
use AmirHossein5\LaravelImage\Facades\Image;

if(!Image::rm($post->image)){
  return back()->with('fail message');
}

$post->delete;
```

```php
use AmirHossein5\LaravelImage\Facades\Image;

$image = Image::setDefaultSizeFor($post->image, $request['default_size']);

if (!$image) {
  return back()
    ->withInput()
    ->withErrors(['image' => __('validation.uploaded')]);
}

$post->image = $image;
$post->save();
```


```php
use AmirHossein5\LaravelImage\Facades\Image;

$request['icon'] = Image::raw($request['icon'])
    ->be('icon.png')
    ->replace(false, function ($image) {
        return $image->imagePath;
    });
    
if (!$request['icon']) {
  return back()
    ->withInput()
    ->withErrors(['image' => __('validation.uploaded')]);
}
```

```php
use AmirHossein5\LaravelImage\Facades\Image;

$img = Intervention::make('https://avatars.githubusercontent.com/u/68776630?s=40&v=4');

$avatar = Image::raw($img)
    ->in('avatar')
    ->save();
```

<br/>

## License

[License](LICENSE)



CHANGELOG.md
# Release Notes

## [v2.2.0](https://github.com/amirHossein5/laravel-image/compare/v2.1.1...v2.2.0)
### Added:
- Transactions [7af08faa](https://github.com/amirHossein5/laravel-image/commit/7af08faa6121f247fe585cb2436472826a0937cb)
- Determining image quality [6cbe253eb](https://github.com/amirHossein5/laravel-image/commit/6cbe253eb0e629103b9ea36b3f851d65842f0dc2)
- \Intervention\Image\Image is acceptable for input of ```raw``` and ```make``` method [6304c3cae](https://github.com/amirHossein5/laravel-image/commit/6304c3caeacc23340f3c93eb82becaa064361510)

### Changed:
- ```in``` method is optional
- ```be``` method improved


## [v2.1.1](https://github.com/amirHossein5/laravel-image/compare/v2.1.0...v2.1.1)

### Fixed
Fix throwing exception when removing a false path (unlocated) image(s) ([2afca87bde](https://github.com/amirHossein5/laravel-image/commit/2afca87bde1882b7ff3f4342f2a58bcd975655c5))


## [v2.1.0](https://github.com/amirHossein5/laravel-image/compare/v2.0.2...v2.1.0)

laravel 9 support


## [v2.0.2](https://github.com/amirHossein5/laravel-image/compare/v2.0.1...v2.0.2)

### Fixed
- On removing image, only if directory is empty, it'll be remove [97d736a](https://github.com/amirHossein5/laravel-image/commit/97d736a2e3f7354ef6ac0bcbd7d1a5622daa457b)


## [v2.0.1](https://github.com/amirHossein5/laravel-image/compare/v2.0.0...v2.0.1)

Set public disk for ```rm``` when not setted.


## [v2.0.0](https://github.com/amirHossein5/laravel-image/compare/v1.2.0...v2.0.0)

### Added
- Disks (saving in storage).

### Changed
 - ```inPath``` to ```in```
 - Result array.
 - Getting result array manually properties.

### Removed
- ```inPublicPath```

tests/TestCase.php
<?php

namespace AmirHossein5\LaravelImage\Tests;

use AmirHossein5\LaravelImage\ImageServiceProvider;
use Illuminate\Http\UploadedFile;
use Illuminate\Support\Facades\File;
use Intervention\Image\Facades\Image as Intervention;
use Orchestra\Testbench\TestCase as TestbenchTestCase;

class TestCase extends TestbenchTestCase
{
    protected UploadedFile $image;

    protected function disk_path(string $disk, string $path): string
    {
        return config('image.disks.'.$disk).DIRECTORY_SEPARATOR.$path;
    }

    protected function getPackageProviders($app)
    {
        return [
            ImageServiceProvider::class,
        ];
    }

    protected function setUp(): void
    {
        parent::setUp();

        $this->image = UploadedFile::fake()->image('test.png');
    }

    protected function tearDown(): void
    {
        if (file_exists(public_path('images'))) {
            File::deleteDirectory(public_path('images'));
            $this->assertFalse(file_exists(public_path('images')));
        }

        parent::tearDown();
    }

    protected function random(bool $hasSuffix = true, string $suffix = null): string
    {
        if ($hasSuffix) {
            $suffix = $suffix ? '_'.$suffix : '_'.rand(100, 999);
        }

        return time().$suffix;
    }

    public function interventionCircle(): \Intervention\Image\Image
    {
        // create empty canvas with background color
        $img = Intervention::canvas(300, 200, '#ddd');

        // draw a filled blue circle
        $img->circle(100, 50, 50, function ($draw) {
            $draw->background('#0000ff');
        });

        // draw a filled blue circle with red border
        $img->circle(10, 100, 100, function ($draw) {
            $draw->background('#0000ff');
            $draw->border(1, '#f00');
        });

        // draw an empty circle with 5px border
        $img->circle(70, 150, 100, function ($draw) {
            $draw->border(5, '000000');
        });

        return $img;
    }

    /**
     * Re modifies string by directory separator.
     *
     * @param string $string
     *
     * @return string
     */
    public function directorySeparator(string $string): string
    {
        $string = str_replace('/', DIRECTORY_SEPARATOR, $string);

        return str_replace('\\', DIRECTORY_SEPARATOR, $string);
    }
}


tests/Feature/BaseTest.php
<?php

namespace AmirHossein5\LaravelImage\Tests\Feature;

use AmirHossein5\LaravelImage\Facades\Image;
use AmirHossein5\LaravelImage\Tests\TestCase;
use Intervention\Image\Facades\Image as Intervention;

class BaseTest extends TestCase
{
    public function test_converts_to_correct_directory_separator()
    {
        Image::fake();

        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->setArchiveDirectories('1\\2\\3\\')
            ->save();

        $this->assertStringContainsString(
            '1'.DIRECTORY_SEPARATOR.'2'.DIRECTORY_SEPARATOR.'3'.DIRECTORY_SEPARATOR,
            $image['imageDirectory']
        );
    }

    public function test_image_that_defined_with_intervention_saves()
    {
        $img = $this->interventionCircle();
        $image = Image::raw($img)
            ->in('circle')
            ->save();

        $this->assertTrue(file_exists(public_path($image['index'])));

        $img = $this->interventionCircle();
        $image = Image::make($img)
            ->setExclusiveDirectory('circle')
            ->save();

        foreach ($image['index'] as $path) {
            $this->assertTrue(file_exists(public_path($path)));
        }

        $img = Intervention::make('https://avatars.githubusercontent.com/u/68776630?s=40&v=4');
        $image = Image::raw($img)
            ->in('avatar')
            ->save();

        $this->assertTrue(file_exists(public_path($image['index'])));
    }

    public function test_raw_creates_image_in_root_public_directory()
    {
        $image = Image::raw($this->image)
            ->in('')
            ->be('root.jpg')
            ->save(false, fn ($image) => $image->imagePath);

        $this->assertFileExists(public_path($image));

        $this->assertTrue(Image::rm($image));

        $image = Image::raw($this->image)->save(false, fn ($image) => $image->imagePath);

        $this->assertFileExists(public_path($image));
    }

    public function test_result_array()
    {
        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->save();

        $this->assertEquals(3, count($image));
        $this->assertArrayNotHasKey('default_size', $image);
        $this->assertDirectoryExists(public_path($image['imageDirectory']));
        $this->assertEquals('public', $image['disk']);

        foreach ($image['index'] as $img) {
            $this->assertFileExists(public_path($img));
        }
        foreach (config('image.'.config('image.use_size')) as $sizeName => $size) {
            $this->assertArrayHasKey($sizeName, $image['index']);
        }

        config(['image.default_size' => 'small']);

        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->disk('storage')
            ->save();

        $this->assertEquals('storage', $image['disk']);
        $this->assertEquals(4, count($image));
        $this->assertEquals($image['default_size'], 'small');

        // raw method
        $image = Image::raw($this->image)->disk('storage')->save();

        $this->assertEquals(3, count($image));
        $this->assertTrue(!is_array($image['index']));
        $this->assertArrayNotHasKey('default_size', $image);
        $this->assertDirectoryExists(storage_path('app/'.$image['imageDirectory']));
        $this->assertEquals('storage', $image['disk']);
        $this->assertFileExists(storage_path('app/'.$image['index']));

        $image = Image::raw($this->image)->resize(50, 30)->save();

        $this->assertEquals(3, count($image));
        $this->assertTrue(!is_array($image['index']));
        $this->assertArrayNotHasKey('default_size', $image);
        $this->assertDirectoryExists(public_path($image['imageDirectory']));
        $this->assertEquals('public', $image['disk']);
        $this->assertFileExists(public_path($image['index']));

        $image = Image::raw($this->image)->resizeBy(config('image.imageSizes'))->save();

        $this->assertEquals(3, count($image));
        $this->assertTrue(is_array($image['index']));
        $this->assertArrayNotHasKey('default_size', $image);
        $this->assertDirectoryExists(public_path($image['imageDirectory']));
        $this->assertEquals('public', $image['disk']);
        foreach (config('image.'.config('image.use_size')) as $sizeName => $size) {
            $this->assertArrayHasKey($sizeName, $image['index']);
        }
        foreach ($image['index'] as $img) {
            $this->assertFileExists(public_path($img));
        }

        config(['image.default_size' => 'small']);

        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->save();

        $this->assertTrue(array_key_exists('default_size', $image));

        $image = Image::raw($this->image)
            ->in('post')
            ->save();

        $this->assertTrue(array_key_exists('index', $image));
        $this->assertTrue(array_key_exists('imageDirectory', $image));
        $this->assertTrue(!array_key_exists('default_size', $image));
    }

    public function test_result_array_can_be_get_manually()
    {
        Image::fake();

        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->save(false, function ($image) {
                return [
                    'index'          => $image->imagePath,
                    'imageDirectory' => $image->imageDirectory,
                ];
            });

        $this->assertEquals(
            $this->directorySeparator(
                config('image.root_directory').'/'.'post'.'/'.date('Y').'/'.date('m').'/'.date('d').'/'.$this->random(false)
            ),
            $image['imageDirectory']
        );

        foreach (config('image.'.config('image.use_size')) as $sizeName => $size) {
            $this->assertArrayHasKey($sizeName, $image['index']);
        }
    }

    public function test_saves_and_replaces_image()
    {
        $image = Image::raw($this->image)
            ->in('')
            ->be('logo.png')
            ->replace(false, function ($image) {
                return $image->imagePath;
            });

        $this->assertFileExists(public_path($image));

        $image = Image::raw($this->image)
            ->be('logo.png')
            ->replace(false, function ($image) {
                return $image->imagePath;
            });

        $this->assertTrue(Image::wasRecentlyRemoved());
        $this->assertTrue($image == true);
        $this->assertFileExists(public_path($image));

        // make method
        $images = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->setImageName('test')
            ->setSizesDirectory('test')
            ->replace(false, function ($image) {
                return $image->imagePath;
            });

        foreach ($images as $image) {
            $this->assertFileExists(public_path($image));
        }

        $images = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->setImageName('test')
            ->setSizesDirectory('test')
            ->replace(false, function ($image) {
                return $image->imagePath;
            });

        $this->assertTrue($images == true);
        $this->assertTrue(Image::wasRecentlyRemoved());

        foreach ($images as $image) {
            $this->assertFileExists(public_path($image));
        }
    }

    public function test_manually_way_properties()
    {
        $random = $this->random(false);

        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->setSizesDirectory($random)
            ->quality(0)
            ->save(false, function ($image) {
                return [
                    'image'              => $image->image,
                    'sizes'              => $image->sizes,
                    'defaultSize'        => $image->defaultSize,
                    'imagePath'          => $image->imagePath,
                    'imageDirectory'     => $image->imageDirectory,
                    'imageName'          => $image->imageName,
                    'imageFormat'        => $image->imageFormat,
                    'disk'               => $image->disk,
                    'rootDirectory'      => $image->rootDirectory,
                    'exclusiveDirectory' => $image->exclusiveDirectory,
                    'archiveDirectories' => $image->archiveDirectories,
                    'sizesDirectory'     => $image->sizesDirectory,
                    'quality'            => $image->quality,
                ];
            });

        $this->propertiesAreValid($image, config('image.imageSizes'), false, 0, $random);

        $image = Image::raw($this->image)
            ->setSizesDirectory($random)
            ->save(false, function ($image) {
                return [
                    'image'              => $image->image,
                    'sizes'              => $image->sizes,
                    'defaultSize'        => $image->defaultSize,
                    'imagePath'          => $image->imagePath,
                    'imageDirectory'     => $image->imageDirectory,
                    'imageName'          => $image->imageName,
                    'imageFormat'        => $image->imageFormat,
                    'disk'               => $image->disk,
                    'rootDirectory'      => $image->rootDirectory,
                    'exclusiveDirectory' => $image->exclusiveDirectory,
                    'archiveDirectories' => $image->archiveDirectories,
                    'sizesDirectory'     => $image->sizesDirectory,
                    'quality'            => $image->quality,
                ];
            });

        $this->propertiesAreValid($image, [], true, 90, $random);

        $image = Image::raw($this->image)
            ->quality(20)
            ->setSizesDirectory($random)
            ->save(false, function ($image) {
                return [
                    'image'              => $image->image,
                    'sizes'              => $image->sizes,
                    'defaultSize'        => $image->defaultSize,
                    'imagePath'          => $image->imagePath,
                    'imageDirectory'     => $image->imageDirectory,
                    'imageName'          => $image->imageName,
                    'imageFormat'        => $image->imageFormat,
                    'disk'               => $image->disk,
                    'rootDirectory'      => $image->rootDirectory,
                    'exclusiveDirectory' => $image->exclusiveDirectory,
                    'archiveDirectories' => $image->archiveDirectories,
                    'sizesDirectory'     => $image->sizesDirectory,
                    'quality'            => $image->quality,
                ];
            });

        $this->propertiesAreValid($image, [], true, 20, $random);

        $image = Image::raw($this->image)
            ->resizeBy(config('image.imageSizes'))
            ->setSizesDirectory($random)
            ->quality(0)
            ->in('post/')
            ->save(false, function ($image) {
                return [
                    'image'              => $image->image,
                    'sizes'              => $image->sizes,
                    'defaultSize'        => $image->defaultSize,
                    'imagePath'          => $image->imagePath,
                    'imageDirectory'     => $image->imageDirectory,
                    'imageName'          => $image->imageName,
                    'imageFormat'        => $image->imageFormat,
                    'disk'               => $image->disk,
                    'rootDirectory'      => $image->rootDirectory,
                    'exclusiveDirectory' => $image->exclusiveDirectory,
                    'archiveDirectories' => $image->archiveDirectories,
                    'sizesDirectory'     => $image->sizesDirectory,
                    'quality'            => $image->quality,
                ];
            });

        $this->propertiesAreValid($image, config('image.imageSizes'), true, 0, $random);
    }

    /**
     * Tests the output of manually given image.
     *
     * @param array $image
     * @param array $sizes
     * @param bool  $isRaw
     * @param int   $quality
     *
     * @return void
     */
    private function propertiesAreValid(array $image, array $sizes = [], bool $isRaw = false, int $quality = 90, int $random): void
    {
        $this->assertEquals(get_class($image['image']), \Intervention\Image\Image::class);
        if ($sizes) {
            foreach ($sizes as $sizeName => $dimensions) {
                $this->assertArrayHasKey($sizeName, $image['sizes']);
            }
        } else {
            $this->assertNull($image['sizes']);
        }
        $this->assertNull($image['defaultSize']);
        if (is_array($image['imagePath'])) {
            foreach ($image['imagePath'] as $path) {
                $this->assertFileExists(public_path($path));
            }
        } else {
            $this->assertFileExists(public_path($image['imagePath']));
        }
        $this->assertDirectoryExists(public_path($image['imageDirectory']));
        $this->assertIsString($image['imageDirectory']);
        if ($sizes) {
            $fileName = pathinfo(public_path($image['imagePath']['small']), PATHINFO_FILENAME);
            $extension = pathinfo(public_path($image['imagePath']['small']), PATHINFO_EXTENSION);
            $this->assertEquals($fileName, $image['imageName'].'_small');
        } else {
            $fileName = pathinfo(public_path($image['imagePath']), PATHINFO_FILENAME);
            $extension = pathinfo(public_path($image['imagePath']), PATHINFO_EXTENSION);
            $this->assertEquals($fileName, $image['imageName']);
        }
        $this->assertEquals($extension, $image['imageFormat']);
        $this->assertEquals('public', $image['disk']);
        if (!$isRaw) {
            $this->assertEquals(config('image.root_directory'), $image['rootDirectory']);
            $this->assertEquals('post', $image['exclusiveDirectory']);
            $this->assertEquals(
                $this->directorySeparator(date('Y').'/'.date('m').'/'.date('d')),
                $image['archiveDirectories']
            );
        }
        $this->assertEquals($random, $image['sizesDirectory']);
        $this->assertEquals($quality, $image['quality']);
    }
}


tests/Feature/FakeTest.php
<?php

namespace AmirHossein5\LaravelImage\Tests\Feature;

use AmirHossein5\LaravelImage\Facades\Image;
use AmirHossein5\LaravelImage\Tests\TestCase;

class FakeTest extends TestCase
{
    public function test_fake_method()
    {
        Image::fake();

        $image = Image::raw($this->image)->save();
        $this->assertFalse(file_exists(public_path($image['index'])));

        $image = Image::raw($this->image)->replace();
        $this->assertFalse(file_exists(public_path($image['index'])));

        $image = Image::make($this->image)
            ->setExclusiveDirectory('post')
            ->save();
        foreach ($image['index'] as $image) {
            $this->assertFalse(file_exists(public_path($image)));
        }
    }
}


config/image.php
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Image Driver
    |--------------------------------------------------------------------------
    |
    | Intervention Image supports "GD Library" and "Imagick" to process images
    | internally. You may choose one of them according to your PHP
    | configuration. By default PHP's "GD Library" implementation is used.
    |
    | Supported: "gd", "imagick"
    |
    */

    'driver' => 'gd',

    /**
     * Disks of paths.
     */
    'disks' => [
        'public'  => public_path(),
        'storage' => storage_path('app'),
    ],

    /**
     * The root directory of all of the images.
     * if you use "make" method you can modify it by "->setRootDirectory('directory name')" method.
     */
    'root_directory' => 'images',

    /**
     * Key of your sizes array, that you defined in below section, like default one.
     * By default the image sizes in "make" method will be create by this returned array.
     */
    'use_size' => 'imageSizes',

    /**
     * It's for making images with these sizes.
     * Add one of these to above section("use_size").
     */
    'imageSizes' => [
        'large' => [
            'width'  => '800',
            'height' => '600',
        ],
        'medium' => [
            'width'  => '400',
            'height' => '300',
        ],
        'small' => [
            'width'  => '80',
            'height' => '60',
        ],
    ],

    /**
     * If you want to use default_size functionality, or not keep it null.
     * Write key of one of your sizes, that have been added in "use_size" section.
     * When you create image,
     * (if "default_size" not be null, and if you have more than one sizes),
     * it returns an array, which contains "default_size" key:
     *     [
     *         'index' => [
     *             'large' => '. . ./1638611107_960_large.png'
     *             'medium' => '. . ./1638611107_960_medium.png'
     *             'small' => '. . ./1638611107_960_small.png'
     *         ],
     *         'imageDirectory' => '. . .',
     *         'default_size' => 'medium'
     *      ].
     */
    'default_size' => null,
];
